pkg/blobinfocache/boltdb/boltdb.go

```diff
@@ -328,57 +345,72 @@ func (bdc *cache) CandidateLocations2(transport types.ImageTransport, scope type
+							res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, d, v2Output)
+						}
+						return nil
+					}); err != nil {
+						return nil, err
+					}
+				}
+			}
+			if uncompressedDigestValue != primaryDigest {
+				res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, uncompressedDigestValue, v2Output)
+			}
+		}
+		return res, nil
+	}
+
 	if err := bdc.view(func(tx *bolt.Tx) error {
 		scopeBucket := tx.Bucket(knownLocationsBucket)
-		if scopeBucket == nil {
-			return nil
+		if scopeBucket != nil {
+			scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
 		}
-		scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
-		if scopeBucket == nil {
-			return nil
-		}
-		scopeBucket = scopeBucket.Bucket([]byte(scope.Opaque))
-		if scopeBucket == nil {
-			return nil
+		if scopeBucket != nil {
+			scopeBucket = scopeBucket.Bucket([]byte(scope.Opaque))
 		}
 		// compressionBucket won't have been created if previous writers never recorded info about compression,
 		// and we don't want to fail just because of that
 		compressionBucket := tx.Bucket(digestCompressorBucket)
 
-		res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, primaryDigest, requireCompressionInfo)
-		if canSubstitute {
-			if uncompressedDigestValue = bdc.uncompressedDigest(tx, primaryDigest); uncompressedDigestValue != "" {
-				b := tx.Bucket(digestByUncompressedBucket)
-				if b != nil {
-					b = b.Bucket([]byte(uncompressedDigestValue.String()))
-					if b != nil {
-						if err := b.ForEach(func(k, _ []byte) error {
-							d, err := digest.Parse(string(k))
-							if err != nil {
-								return err
-							}
-							if d != primaryDigest && d != uncompressedDigestValue {
-								res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, d, requireCompressionInfo)
-							}
-							return nil
-						}); err != nil {
-							return err
-						}
-					}
-				}
-				if uncompressedDigestValue != primaryDigest {
-					res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, uncompressedDigestValue, requireCompressionInfo)
+		// Add alls potential blobs with location unknown if v2Output is configured true.
+		resAllBlobs, err = getCandidatesForUncompressedDigest(tx, resAllBlobs, nil, compressionBucket, primaryDigest, v2Output)
+		if err != nil {
+			return err
+		}
+
+		if scopeBucket != nil {
+			res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, primaryDigest, v2Output)
+			if canSubstitute {
+				res, err = getCandidatesForUncompressedDigest(tx, res, scopeBucket, compressionBucket, primaryDigest, v2Output)
+				if err != nil {
+					return err
 				}
 			}
 		}
 		return nil
 	}); err != nil { // Including os.IsNotExist(err)
 		return []blobinfocache.BICReplacementCandidate2{} // FIXME? Log err (but throttle the log volume on repeated accesses)?
 	}
-
+	res = append(res, resAllBlobs...)
```

Comment:
If this is just going to collect everything into `res`, the code can just be appending to `res` throughout, without needing an extra `resAllBlobs` variable.

But see elsewhere.